hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageDispatcherService.java

```diff
@@ -500,4 +515,55 @@ private List<SoftwareModuleMetadata> getSoftwareModuleMetadata(final SoftwareMod
                 PageRequest.of(0, RepositoryConstants.MAX_META_DATA_COUNT), module.getId()).getContent();
     }
 
+    private void sendUpdateMessageToTargets(final Map<String, ActionProperties> actions, final List<Target> targets,
+                                            final Map<SoftwareModule, List<SoftwareModuleMetadata>> modules) {
+
+        List<DmfTarget> dmfTargets = targets.stream().filter(target -> IpUtil.isAmqpUri(target.getAddress()))
+                .map(t -> convertToDmfTarget(t, actions.get(t.getControllerId()).getId())).collect(Collectors.toList());
+
+        final DmfBatchDownloadAndUpdateRequest batchRequest = new DmfBatchDownloadAndUpdateRequest();
+        batchRequest.setTimestamp(System.currentTimeMillis());
+        batchRequest.setTargets(dmfTargets);
+
+        if (modules != null) {
+            modules.entrySet().forEach(entry ->
+                    batchRequest.addSoftwareModule(convertToAmqpSoftwareModule(targets.get(0), entry)));
+        }
+
+        Optional<ActionProperties> first = actions.values().stream().findFirst();
+
+        final Message message = getMessageConverter().toMessage(batchRequest,
+                    createMessagePropertiesBatch(first.get().getTenant(), getBatchEventTopicForAction(first.get())));
+            amqpSenderService.sendMessage(message, targets.get(0).getAddress());
+    }
+
+    protected DmfTarget convertToDmfTarget(final Target target, final Long actionId) {
+        DmfTarget dmfTarget = new DmfTarget();
+        dmfTarget.setActionId(actionId);
+        dmfTarget.setControllerId(target.getControllerId());
+        dmfTarget.setTargetSecurityToken(systemSecurityContext.runAsSystem(target::getSecurityToken));
+        return dmfTarget;
+    }
+
+    private static MessageProperties createMessagePropertiesBatch(final String tenant, final EventTopic topic) {
+        final MessageProperties messageProperties = new MessageProperties();
+        messageProperties.setContentType(MessageProperties.CONTENT_TYPE_JSON);
+        messageProperties.setHeader(MessageHeaderKey.CONTENT_TYPE, MessageProperties.CONTENT_TYPE_JSON);
+        messageProperties.setHeader(MessageHeaderKey.TENANT, tenant);
+
+        messageProperties.setHeader(MessageHeaderKey.TOPIC, topic);
+        messageProperties.setHeader(MessageHeaderKey.TYPE, MessageType.EVENT);
+        return messageProperties;
+    }
+
+    public boolean isBatchAssignmentsEnabled() {
+        return systemSecurityContext.runAsSystem(() -> tenantConfigurationManagement
+                .getConfigurationValue(BATCH_ASSIGNMENTS_ENABLED, Boolean.class).getValue());
+    }
+
+    private static EventTopic getBatchEventTopicForAction(final ActionProperties action) {
+        return (Action.ActionType.DOWNLOAD_ONLY == action.getActionType() || !action.isMaintenanceWindowAvailable())
```

Comment:
Should the `action.isMaintenanceWindowAvailable()` be really negated here? With current logic the topic would be `BATCH_DOWNLOAD` for all actions without maintenance window, shouldn't it be other way around, or am I missing something?

Reply:
ok, `isMaintenanceWindowAvailable` method indeed checks if maintenance takes place in the time of calling it. I misinterpreted it, assuming it only checks that maintenance window option was been selected that is done by `hasMaintenanceSchedule`. Nevertheless, I think method naming is misleading and we should consider changing it someday